src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs

```diff
@@ -0,0 +1,293 @@
+        {
+            // Allow inlining if the caller is within the current version bubble
+            // (because otherwise we may not be able to encode its tokens)
+            // and if the callee is either in the same version bubble or is marked as non-versionable.
+            bool canInline = VersionsWithMethodBody(callerMethod) &&
+                (VersionsWithMethodBody(calleeMethod) ||
+                    calleeMethod.HasCustomAttribute("System.Runtime.Versioning", "NonVersionableAttribute"));
+
+            return canInline;
+        }
+
+        public sealed override bool GeneratesPInvoke(MethodDesc method)
+        {
+            // PInvokes depend on details of the core library, so for now only compile them if:
+            //    1) We're compiling the core library module, or
+            //    2) We're compiling any module, and no marshalling is needed
+            //
+            // TODO Future: consider compiling PInvokes with complex marshalling in version bubble
+            // mode when the core library is included in the bubble.
+
+            Debug.Assert(method is EcmaMethod);
+
+            // If the PInvoke is declared on an external module, we can only compile it if
+            // that module is part of the version bubble.
+            if (!_versionBubbleModuleSet.Contains(((EcmaMethod)method).Module))
+                return false;
+
+            if (((EcmaMethod)method).Module.Equals(method.Context.SystemModule))
+                return true;
+
+            return !Marshaller.IsMarshallingRequired(method);
+        }
+
+        public sealed override bool TryGetModuleTokenForExternalType(TypeDesc type, out ModuleToken token)
+        {
+            Debug.Assert(!VersionsWithType(type));
+
+            if (_typeRefsInCompilationModuleSet == null)
+            {
+                _typeRefsInCompilationModuleSet = new Dictionary<TypeDesc, ModuleToken>();
+
+                foreach (var module in _compilationModuleSet)
+                {
+                    EcmaModule ecmaModule = (EcmaModule)module;
+                    foreach (var typeRefHandle in ecmaModule.MetadataReader.TypeReferences)
+                    {
+                        try
+                        {
+                            TypeDesc typeFromTypeRef = ecmaModule.GetType(typeRefHandle);
+                            if (!_typeRefsInCompilationModuleSet.ContainsKey(typeFromTypeRef))
+                            {
+                                _typeRefsInCompilationModuleSet.Add(typeFromTypeRef, new ModuleToken(ecmaModule, typeRefHandle));
+                            }
+                        }
+                        catch (TypeSystemException) { }
+                    }
+                }
+            }
+
+            return _typeRefsInCompilationModuleSet.TryGetValue(type, out token);
+        }
+
+        private bool ComputeTypeVersionsWithCode(TypeDesc type)
+        {
+            if (type == type.Context.CanonType)
```

Comment:
Should we use `IsCanonicalDefinitionType` to also cover other types of canon we might add in the future (universal, enum integer, etc.)?

Reply:
Yes. This is a holdover from some debugging I was doing that led me to discover we were loading System.__Canon as an EcmaType as well as a CanonType.